runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -112,10 +115,61 @@ public void translate(Pipeline pipeline) {
                         flinkRunner)))
             .build();
 
+    generateDummyOperatorForLonelySource(pipeline);
     pipeline.replaceAll(transformOverrides);
     super.translate(pipeline);
   }
 
+  /**
+   * StreamExecutionEnvironment does not allow a lonely Source to execute.
+   * So generate a dummy operator after the source if we only have lonely source.
+   */
+  private void generateDummyOperatorForLonelySource(Pipeline pipeline) {
+
+    final Map<PValue, Boolean> sources = new HashMap<>();
+
+    pipeline.traverseTopologically(new Pipeline.PipelineVisitor.Defaults() {
+
+      @Override
+      public void visitPrimitiveTransform(TransformHierarchy.Node node) {
+
+        // find source
+        if (!node.isRootNode()) {
+          if (node.getInputs().size() == 0) {
+            for (PValue value : node.getOutputs().values()) {
+              sources.put(value, false);
+            }
+          }
+        }
+
+        // update source not lonely
+        if (!node.isRootNode()) {
+          for (PValue value : node.getInputs().values()) {
+            if (sources.containsKey(value)) {
+              sources.put(value, true);
+            }
+          }
+        }
+      }
+    });
+
+    // add dummy parDo
+    for (Map.Entry<PValue, Boolean> entry : sources.entrySet()) {
+      if (!entry.getValue()) {
+        PCollection<?> pCollection = (PCollection) entry.getKey();
+        pCollection.apply(ParDo.of(new EmptyDoFn<>()));
```

Comment:
@JingsongLi This seems to do the trick. 👍 I just want to double check with @tgroh  if there is a obvious canonical way of doing this that we're missing.

@tgroh Do you have any opinion on this, as the master Pipeline surgeon? 😄